client/beefy/src/worker.rs

```diff
@@ -196,132 +215,111 @@ where
+				Ok(h) => h,
+				Err(e) => {
+					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
+						block_num, e);
+					None
+				},
+			};
+			// Update RPC worker with new best BEEFY block hash.
+			block_hash.map(|hash| {
 				self.beefy_best_block_sender
-					.notify(|| Ok::<_, ()>(notification.hash.clone()))
-					.expect("forwards closure result; the closure always returns Ok; qed.");
-
-				// this metric is kind of 'fake'. Best BEEFY block should only be updated once we
-				// have a signed commitment for the block. Remove once the above TODO is done.
-				metric_set!(self, beefy_best_block, *notification.header.number());
-			}
+					.notify(|| Ok::<_, ()>(hash))
+					.expect("forwards closure result; the closure always returns Ok; qed.")
+			});
+			// Set new best BEEFY block number.
+			self.best_beefy_block = Some(block_num);
+			metric_set!(self, beefy_best_block, block_num);
+		} else {
+			debug!(target: "beefy", "🥩 Can't set best beefy to older: {}", block_num);
 		}
+	}
 
-		if self.should_vote_on(*notification.header.number()) {
-			let (validators, validator_set_id) = if let Some(rounds) = &self.rounds {
-				(rounds.validators(), rounds.validator_set_id())
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator set - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-			let authority_id = if let Some(id) = self.key_store.authority_id(validators) {
-				debug!(target: "beefy", "🥩 Local authority id: {:?}", id);
-				id
-			} else {
-				debug!(target: "beefy", "🥩 Missing validator id - can't vote for: {:?}", notification.header.hash());
-				return
-			};
-
-			let mmr_root =
-				if let Some(hash) = find_mmr_root_digest::<B, Public>(&notification.header) {
-					hash
-				} else {
-					warn!(target: "beefy", "🥩 No MMR root digest found for: {:?}", notification.header.hash());
-					return
-				};
-
-			let payload = Payload::new(known_payload_ids::MMR_ROOT_ID, mmr_root.encode());
-			let commitment = Commitment {
-				payload,
-				block_number: notification.header.number(),
-				validator_set_id,
-			};
-			let encoded_commitment = commitment.encode();
+	/// Handle session changes by starting new voting round for mandatory blocks.
+	fn init_session_at(&mut self, active: ValidatorSet<AuthorityId>, session_start: NumberFor<B>) {
+		debug!(target: "beefy", "🥩 New active validator set: {:?}", active);
+		metric_set!(self, beefy_validator_set_id, active.id());
+		// BEEFY should produce a signed commitment for each session
+		if active.id() != self.last_signed_id + 1 && active.id() != GENESIS_AUTHORITY_SET_ID {
+			metric_inc!(self, beefy_skipped_sessions);
+		}
 
-			let signature = match self.key_store.sign(&authority_id, &*encoded_commitment) {
-				Ok(sig) => sig,
-				Err(err) => {
-					warn!(target: "beefy", "🥩 Error signing commitment: {:?}", err);
-					return
-				},
-			};
+		if log_enabled!(target: "beefy", log::Level::Debug) {
```

Comment:
I would do it regardless. Probably `rustc` is smart enough to generate code that is similar to this.